src/main/kotlin/model/LotteryPurchasePattern.kt

```diff
@@ -0,0 +1,35 @@
+package model
+
+class LotteryPurchasePattern(val manualLottoCount: Int, val autoLottoCount: Int) {
```

Comment:
```suggestion
data class LotteryPurchasePattern(val manualLottoCount: Int, val autoLottoCount: Int) {
```

LotteryPurchasePattern 이 현재 ofManual  말고는 그 어떤 로직도 가지고 있지 않는 듯 한데,
data class 로 취급해도 되지 않을까 해서 제안해봅니다.

Reply:
구현한 이유를 스스로 돌아볼 수 있는 질문을 주셔서 감사합니다.
저는 기본적으로는 클래스를 사용하면서, 데이터 클래스 적용에는 다음과 같은 점을 고려하는 것 같습니다.

1. 자동으로 생성되는 오버라이드 메서드들이 현재 필요한지
2. 아니면 이 클래스가 단순 데이터만을 보관하고 전달하는데에만 사용되는지

그 누군가가 이야기해주진 않았지만, 저는 의미론적으로 사용하고 있습니다.
데이터의 보관과 전달에 있어서는 보통 데이터 클래스를 사용하고 있고,
데이터와 함께 처리 기능이 있을 때는 보통 클래스를 사용하고 있는 것 같습니다.

그래서 팩토리 메서드와 데이터 처리를 담은 클래스이면서, 오버라이드되는 메서드들도 필요가 없다고 생각하여 클래스로 구현하였습니다.

한없이 부족한 지식에 도움을 주실 수 있다면 정말 감사드리겠습니다. 의견을 남겨주시면 제 생각 확장에 큰 도움이 될 것 같습니다.

Reply:
> 그래서 팩토리 메서드와 데이터 처리를 담은 클래스이면서, 오버라이드되는 메서드들도 필요가 없다고 생각하여 클래스로 구현하였습니다.

이런 기준이 있다면 저는 충분하다고 생각합니다.
저는 특히 팡태가 말씀주신 "이 클래스가 단순 데이터만을 보관하고 전달하는데에만 사용되는지" 조건은 저도 크게 동의하는 조건입니다.

제가 LotteryPurchasePattern 에 대해서 data class 를 제안한 이유는, LotteryPurchasePattern 가 ofManual 말고는 유의미한 메서드가 없어서, 
"데이터만을 보관하고 전달하는데에만 사용" 되고 있는 듯 하였기 때문입니다.

혹시 팡태가 생각하시기에 LotteryPurchasePattern 에 데이터의 보관과 전달 외에 다른 역할과 책임이 있는지 여쭈어봅니다.

Reply:
제 생각에는 유효성 검사를 품고 있다는 점이 조금 애매한 경계였던 것 같습니다.
혹시 이에 대해서는 어떻게 생각하시는지 제가 여쭤볼 수 있을까요?

Reply:
객체는 상태(멤버)와 그에 따른 로직(메서드)를 통해 그 역할과 책임이 나타납니다.
유효성 검사는 현재 객체가 상태를 갖기 전 "생성"할 때 적용되는 로직이고,
객체가 실제로 상태를 가진 후에 갖게 되는 로직은 현재 전혀 없습니다.

상태를 갖기전의 로직은 극단적으로 말해서, 외부 유틸의 static 메서드로 제공되어도 아무 문제가 없습니다.
그렇기에 팡태님도 현재 유효성 검사를 내부 object 의 static 메서드로 제공하고 있습니다.
팡태는 LotteryPurchasePattern 의 역할과 책임이 "유효성 검사" 라고 생각하시나요?

만약 팡태가 의도한 역할과 책임이 유효성 검사가 전부이고, 
그에 따라 LotteryPurchasePattern 라는 객체의 이름을 지으신게 맞다면,
저는 객체 이름에서 유효성 검사라는 역할과 책임을, 동료 개발자의 입장에서 유추하기 어려웠다고 피드백 드려봅니다.

Reply:
어떤 말씀이신지 와닿았습니다. 설명을 너무 잘해주시는 것 같아요.
현재 제가 무언가를 헷갈려하고 있는 것 같은데, 바로잡아볼 부분이 보여지면 좋겠습니다.

저의 개념적인 부분이 부족할 수 있지만, LotteryPurchasePattern 을 처음 설계할 때
만약 사용자가 5천원을 지불했고, '수동 로또 갯수 입력' 이 3이라고 주어진다면,
수동 로또: 3, 자동 로또: 2 라는 결과를 계산해주면서 각 값 또한 멤버로 갖고 있는 '구매 패턴에 대한 책임을 갖는' 클래스를 정의하면 좋겠다 라고 생각했습니다.

유효성 검사가 애매하다고 생각했던 것은, 에러 메세지도 그렇고 'LotteryPurchasePattern' 에 종속되는 유효성 검사라고 생각했기 때문인데요..!

생성 과정에서 1. 결과 계산과 2. 유효성 검사를 진행해준 뒤에 생성합니다. 
이 생성 과정은 클래스의 역할이라고 볼 순 없는 개념인가요??

Reply:
생성 과정도 클래스의 역할중 하나입니다. 하지만 생성과정 "만이" 클래스의 역할로 하려면, 저는 굳이 "클래스" 혹은 객체지향적인 설계가 필요한가? 라고 생각합니다.
그런 방식의 설계는 전통적인 C 의 struct 문법만으로도 충분히 구현할 수 있기 때문입니다.

LotteryPurchasePattern 를 설계를 할 때,  '구매 패턴에 대한 책임을 갖는' 클래스를 정의하려 하셨다고 들었습니다.
저도 클래스명을 보고 구매 패턴에 대한 역할이 LotteryPurchasePattern 에 있을 거라고 기대했습니다.

그리고 여기서의 구매 패턴이란, manualLottoCount 만큼 수동으로 로또를 발행하고, autoLottoCount 만큼 자동으로 로또를 발행한다 입니다.
그렇다면 저는 클래스명만보고 기대하는 것은 단순히 manualLottoCount 와 autoLottoCount 라는 숫자 데이터를 제공하는게 아니라, 오히려 이 숫자 데이터는 캡슐화되어 접근할 수 없고, 실제로 로또를 발행하거나,
로또를 발행하기에는 역할이 너무 커지는 것 같다면, 현재 자동 혹은 수동 로또의 발행이 가능한 상태인지 알려주는 boolean 값을 계산해서 보여주지 않을까 기대했습니다.
그렇다면 LotteryPurchasePattern 는 단순히 숫자 데이터를 제공하는 객체가 아니라, 역할과 책임에 걸맞는 유의미한 로직을 제공하는 객체가 됩니다.

지금은 역할과 책임에 걸맞는 유의미한 로직이 전혀 보이질 않습니다.

Reply:
@ghojeong 제가 바라보지 못한 넓은 관점의 견해를 남겨주셔서 감사드립니다. 말씀하신 내용을 토대로 스스로 생각해보니, 현재 LotteryPurchasePattern은 단지 유효성 검증만을 갖고 있고 이는 객체지향적인 설계로써 접목되기에는 아쉽다고 생각했습니다. 해당 클래스는 현재 파이로가 말씀해주신 것 처럼 데이터로써의 관점이 크기 때문에 data class로 정의하는게 더 가깝다는 생각을 할 수 있게 되었네요. 감사합니다.